### PR TITLE
gscan2pdf: fix build failure, add patches

### LINKED ISSUE #268406: Build failure: gscan2pdf
### Steps To Reproduce

1. build `gscan2pdf` (e.g. `nix-build -A gscan2pdf` on a current revision. I used 168ba6d83521c5cd6a8f3375950ec4f4c48ee29a)

### Build log

see https://hydra.nixos.org/build/241329132/nixlog/2

### Additional context

`gscan2pdf` times out on test `t/135_save_tiff_as_ps_with_space.t`

Upstream has no bug report concerning this test.

~~I am bisecting currently to figure out what caused the test to fail.~~

`git bisect --first-parent` returned:

```
# first bad commit: [a7a8f8253db3f32cc15baf527d36ef7171425eba] Merge #263535: staging-next 2023-10-26
```
#263535
The bisect on the actual commit will take a _long_ time, though..

### Notify maintainer

@pacien

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.62, NixOS, 23.11 (Tapir), 23.11.20231112.a49378b`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixos, nixos-unstable"`
 - channels(wogan): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by gador**
## Description of changes

Due to the removal of tools like `tiff2pdf` from `libtiff` many tests fail or cause a timeout.
This commit fixes the dependency to `libtiff_4_5`. Also it includes two upstream patches, which haven't been merged and re-adds a previously removed test.

fixes #268406

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by pacien — APPROVED** [review]
Thank you for handling this!

- [x] diff looks good
- [x] tests passing
- [x] program starting

(I would just split the mutually independent changes into four commits).

**Comment by gador** [comment]
> (I would just split the mutually independent changes into four commits).

Done 👍

**Review by pacien — APPROVED** [review]
Nit: Message of the first commit (e7090c1) should be updated accordingly.

Other than that, all good, thanks 👍

**Comment by gador** [comment]
> Nit: Message of the first commit (e7090c1) should be updated accordingly.

Good catch, done :+1:

**Review by pacien — APPROVED** [review]

File: pkgs/applications/graphics/gscan2pdf/default.nix [hunk 1]
```diff
@@ -17,6 +17,13 @@ perlPackages.buildPerlPackage rec {
     hash = "sha256-NGz6DUa7TdChpgwmD9pcGdvYr3R+Ft3jPPSJpybCW4Q=";
   };
 
+    patches = [
+    # fixes warnings during tests. See https://sourceforge.net/p/gscan2pdf/bugs/421
```
Comment (SuperSandro2000) [inline pkgs/applications/graphics/gscan2pdf/default.nix:21]: This patch is rather large and only fixes warnings. Do we really need it?

File: pkgs/applications/graphics/gscan2pdf/default.nix [hunk 2]
```diff
@@ -87,7 +94,7 @@ perlPackages.buildPerlPackage rec {
 
   nativeCheckInputs = [
     imagemagick
-    libtiff
+    libtiff_4_5 # needs older libtiff version. See commit f57a4b0a for more infos
```
Comment (SuperSandro2000) [inline pkgs/applications/graphics/gscan2pdf/default.nix:97]: The reason should be written here. Also using short Shas for long term is not safe.

File: pkgs/applications/graphics/gscan2pdf/default.nix (hunk 1 above)
Comment (gador) [reply pkgs/applications/graphics/gscan2pdf/default.nix:21]: I switched to `fetchpatch`, so it shouldn't really matter for nixpkgs. 

Really necessary: No, but will probably be integrated in the next update and it makes output (and debugging) easier.

File: pkgs/applications/graphics/gscan2pdf/default.nix (hunk 2 above)
Comment (gador) [reply pkgs/applications/graphics/gscan2pdf/default.nix]: Done

**Review by pacien — APPROVED** [review]

**Pull request merged**